### ISSUE TITLE
Validation error message is empty on array of file

### ISSUE THREAD

**Issue by lino10** [laravel/framework#12118]
Hi, after doing some testing, I think this is a bug.

I have this html
```
<input type="file" name="images[]">
<input type="file" name="images[]">
.
.
<input type="file" name="images[]">
```

Form Request's rules
```
'images.*' => 'mimes:jpeg',
'name' => 'required',
```

When the file type is not jpeg, it redirects back to the input page. So the validation works, but there's no error message. When I checked the error message bag, it is empty.
```
Illuminate\Support\ViewErrorBag Object ( [bags:protected] => Array ( ) ) 
```

I'm using Laravel Framework version 5.2.12

Might be related: https://github.com/laravel/framework/issues/6028

**Comment by GrahamCampbell** [comment]
Thanks for reporting. Ping @taylorotwell.

**Comment by ctf0** [comment]
same here

**Comment by lino10** [comment]
I have more information.

if the file validation trigerred, the 'old' field form data will be gone too.

So if I input 'name' and got wrong validation on the file, the field name will not be filled the next redirect.

**Comment by ctf0** [comment]
@lino10 i believe thats because the image validation bug doesn't treat this as a validation error rather its more of just redirect back.

**Comment by ctf0** [comment]
@lino10 here is a solution i used back in 4.2, give it a try and see if u have any success
```
if (Input::file('thumb')->getError() != 0) {
                return Redirect::back()
                    ->withInput()
                    ->withErrors(Input::file('thumb')->getErrorMessage());
            }
```

**Comment by lino10** [comment]
@ctf0 Hi, I use the Form Request Validation method, how to implement your code using it?
https://laravel.com/docs/5.2/validation#form-request-validation

**Comment by ctf0** [comment]
@lino10 ah yes, unfortunately that method doesn't give much of choice to customize the pipeline as u want however u can make the validation along with the rules in another class and inject it into ur store method and work on it as usual.

**Comment by kkiernan** [comment]
I am not able to recreate this issue. Form request validation on an array of files seems to work fine for me on a clean install of Laravel 5.2.12. Has this been fixed already? Maybe something else is going on?

**Comment by ctf0** [comment]
@kkiernan did that work with the `min|max` options on the array ?

**Comment by kkiernan** [comment]
Yes I get all error messages as expected. For example, if I upload a gif for the first image, this is the error bag I receive:

```
ViewErrorBag {#150 ▼
  #bags: array:1 [▼
    "default" => MessageBag {#151 ▼
      #messages: array:1 [▼
        "images.0" => array:1 [▼
          0 => "The images.0 must be a file of type: jpeg."
        ]
      ]
      #format: ":message"
    }
  ]
}
```

**Comment by ctf0** [comment]
@kkiernan , Ping @GrahamCampbell 
right now am on **5.2.14** and i have the rules as
`'photo.*' => 'required|image|min:1|max:4000|mimes:jpeg,bmp,png'`
and to test i've chose a `gif` file with size of `4.3mb`,
so i should get 2 validation errors for both the type and size, however am only getting

```
Validator {#252 ▼
  #translator: Translator {#114 ▶}
  #presenceVerifier: DatabasePresenceVerifier {#141 ▶}
  #container: Application {#2 ▶}
  #failedRules: array:1 [▼
    "photo.0" => array:1 [▼
      "Required" => []
    ]
  ]
  #messages: MessageBag {#253 ▼
    #messages: array:1 [▼
      "photo.0" => array:1 [▼
        0 => "The photo.0 field is required."
      ]
    ]
    #format: ":message"
  }
  #data: array:14 [▶]
  #files: array:1 [▼
    "photo.0" => UploadedFile {#30 ▼
      -test: false
      -originalName: "3o85xpXcGqIX5ALSlG.gif"
      -mimeType: "application/octet-stream"
      -size: 0
      -error: 1
      path: ""
      filename: ""
      basename: ""
      pathname: ""
      extension: ""
      realPath: "/Users/Novo/Public/www/public"
      aTime: 1970-01-01 00:00:00
      mTime: 1970-01-01 00:00:00
      cTime: 1970-01-01 00:00:00
      inode: false
      size: false
      perms: 00
      owner: false
      group: false
      type: false
      writable: false
      readable: false
      executable: false
      file: false
      dir: false
      link: false
    }
  ]
  #rules: array:1 [▼
    "photo.0" => array:5 [▼
      0 => "required"
      1 => "image"
      2 => "min:1"
      3 => "max:4000"
      4 => "mimes:jpeg,bmp,png"
    ]
  ]
  #after: []
  #customMessages: []
  #fallbackMessages: array:2 [▶]
  #customAttributes: []
  #customValues: []
  #extensions: array:2 [▶]
  #replacers: []
  #sizeRules: array:4 [▶]
  #numericRules: array:2 [▶]
  #implicitRules: array:8 [▶]
}
```

**Comment by lino10** [comment]
I just tested using the fresh 5.2.14, there's still no error messages. So why is this issue closed?

Here's the app and resources folder, you just need to replace them to test.
http://s000.tinyupload.com/index.php?file_id=71431977644503943832

Tested using xampp and homestead

**Comment by lino10** [comment]
I think I know why the error displayed for kkiernan. If you don't use enctype="multipart/form-data" on the form, the error will display.

**Comment by kkiernan** [comment]
@lino10 No, my form does have the enctype set.

**Comment by kkiernan** [comment]
@lino10 I just downloaded your files and they work fine for me. Hmm.

![image](https://cloud.githubusercontent.com/assets/3057556/12824757/cccee3b0-cb40-11e5-839f-da05a090eabe.png)

**Comment by kkiernan** [comment]
@lino10 @ctf0 Ok, I am able to recreate this issue now. I had to create a new Homestead VM, and on that box I see the same thing you guys see. Weird. The Illuminate\Support\ViewErrorBag just comes back with an empty bags array. If I run the same example from `php artisan serve` it works fine. How about you guys?

**Comment by ctf0** [comment]
am having this issue through `php artisan serve` on local server not the vm, and i don't know what exactly should i be looking for or how to exactly debug that :(

**Comment by kkiernan** [comment]
@ctf0 The MessageBag actually looks good all the way through the HttpResponseException is thrown in the failedValidation method of the Illuminate\Foundation\Http\FormRequest class. I'm not sure what handles this thought. Below is the error that is logged. Do you get a similar error in your storage/logs directory?

```
[2016-02-04 21:40:45] local.ERROR: Exception: Serialization of 'Symfony\Component\HttpFoundation\File\UploadedFile' is not allowed in /home/vagrant/Code/testing/vendor/laravel/framework/src/Illuminate/Session/Store.php:262
Stack trace:
#0 /home/vagrant/Code/testing/vendor/laravel/framework/src/Illuminate/Session/Store.php(262): serialize(Array)
#1 /home/vagrant/Code/testing/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(88): Illuminate\Session\Store->save()
#2 /home/vagrant/Code/testing/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(155): Illuminate\Session\Middleware\StartSession->terminate(Object(Illuminate\Http\Request), Object(Illuminate\Http\RedirectResponse))
#3 /home/vagrant/Code/testing/public/index.php(65): Illuminate\Foundation\Http\Kernel->terminate(Object(Illuminate\Http\Request), Object(Illuminate\Http\RedirectResponse))
#4 {main}  
```

**Comment by kkiernan** [comment]
Potrential fix: Try replacing the `withInput `method in the `Illuminate\Http\RedirectResponse` class with the one below. I only added a return statement inside the conditional that checks if `$value` is an array. This should prevent it from executing the following line and adding the array to the session which causes the eventual serialization error.

```
    /**
     * Flash an array of input to the session.
     *
     * @param  array  $input
     * @return $this
     */
    public function withInput(array $input = null)
    {
        $input = $input ?: $this->request->input();

        $this->session->flashInput($data = array_filter($input, $callback = function (&$value) use (&$callback) {
            if (is_array($value)) {
                $value = array_filter($value, $callback);
                return;
            }

            return ! $value instanceof UploadedFile;
        }));

        return $this;
    }
```

**Comment by ctf0** [comment]
@kkiernan good find, let me test and come back to u asap.

**Comment by lino10** [comment]
@kkiernan Wow! That fix is working! The 'old' data is also returned successfully! :D

**Comment by kkiernan** [comment]
@lino10 Nice! It's odd that it happens on homestead but not my host machine. I am also on Windows 10 64-bit. Simple fix though. Is it worth opening a pull request for this?

**Comment by ctf0** [comment]
@kkiernan sorry to tell u that nothing changed here, still getting the same old required msg.

also i found another bug which if i uploaded a big file like a 12mb file, i get redirected back with all the inputs are gone and validation errors are showin up for all the fields except the file upload errors.

**Comment by lino10** [comment]
@GrahamCampbell  @taylorotwell 
So is this a confirmed bug or not?
I hope that this can be reopen so it can be fixed..

@kkiernan 
Maybe you can check the array using this?
```
instanceof UploadedFile
```
I'm not expert of laravel framework, just a suggestion

**Comment by kkiernan** [comment]
@lino10 Well the array will never be an instance of UploadedFile. It's more that the array has an UploadedFile object in it that should not be included. At least that is what is happening for me.

@ctf0 Do you have any errors in your logs that you can share?

**Comment by ctf0** [comment]
@kkiernan right now am downloading the vm to see if this is related to my machine or not, about the logging is there something like the `dd()` i can use to get directly to this issue ?

**Comment by kkiernan** [comment]
@ctf0 Can you take a look in storage/logs and see if anything has been logged?

**Comment by ctf0** [comment]
@kkiernan sorry for the long delay, let me do some tests and will come back to u with the results

**Comment by kkiernan** [comment]
@ctf0 No problem. It's such a weird bug. I took another look with fresh eyes and the existing Laravel code seems like it should work totally fine. Really odd.

**Comment by ctf0** [comment]
@kkiernan believe it or not actually this bug is there since 4.2 :cry:

**Comment by lino10** [comment]
@kkiernan @ctf0 
OK, I know how to 100% reproduce this.
Is there any way for @GrahamCampbell @taylorotwell to notice? It seems that after the issue is closed, the notification will not come to them? Do I need to create new issue?

---

So here's to reproduce:
1. Install https://box.scotch.io/
2. Install Laravel
3. Replace the laravel files with http://s000.tinyupload.com/index.php?file_id=71431977644503943832

4. Try uploading the a jpeg/png file. The error will will show that it needs gif image. So that means the bug is not there yet

5. Update to php 7.0 -> https://www.digitalocean.com/community/tutorials/how-to-upgrade-to-php-7-on-ubuntu-14-04
6. Try to do step #4 again, now it will not show error

So my conclusion is, something in php 7.0 break the validation logic

**Comment by lino10** [comment]
@kkiernan 
I just did a bit tinkering, I think this code should fix for the empty array text.

```
    public function withInput(array $input = null)
    {
        $input = $input ?: $this->request->input();

        $this->session->flashInput($data = array_filter($input, $callback = function (&$value) use (&$callback) {
            if (is_array($value)) {
                $value = array_filter($value, $callback);
				
				if (empty($value))
				return false;
            }

            return ! $value instanceof UploadedFile;
        }));

        return $this;
    }
```

**Comment by ctf0** [comment]
okay, i found a solution may need some cleaning , but it works as it should now

1- no change to the `RedirectResponse.php`
2- the validation as follow `'photo.*' => 'required|image|max:2048'`

```php
$validate = Validator::make($request->all(), ['photo.*' => 'required|image|max:2048']);
$photos   = $request->file('photo');

if ($validate->fails()) {

    // check if we have file to avoid an error with getErrorMessage()
    if ($request->hasFile('photo')) {

        // cuz getErrorMessage() doesnt work on array
        foreach ($photos as $photo) {
            if ($photo->getError() != 0) {

                // save an array of uploaded files with errors
                $ph_er[] = $photo->getErrorMessage();
            }
        }
        return back()
            ->withInput()
            ->withErrors($ph_er);
    }
    return back()
        ->withInput()
        ->withErrors($validate);
}
```

- what will happen is all ur validation will work without any issues and if u uploaded an array of images , the posting still fail but now u will get a correct error only for the files that didn't pass otherwise it will save :metal:

**Comment by loursbourg** [comment]
Validator works fine for strings and other stuff, but not for files.
I use the following rules:
`required|max:855|mimes:jpeg,bmp,png`
I'm still facing this issue, please fix it.

**Comment by ctf0** [comment]
@loursbourg validation works fine for single file, the issue is when u try to validate an array

**Comment by KayTokyo** [comment]
Was this issue fixed? Im still having this problem

**Comment by ctf0** [comment]
@KayTokyo open a new issue as this is closed and no one will look into it.

**Comment by vinaysaini89** [comment]
#16113
